File: claude_exp/routers/router_java/src/main/java/com/xv6/shared/Stats.java
package com.xv6.shared;

import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Thread-safe rolling counters over windows [30, 60, 180, 1800] seconds. Port of
 * router_py's shared/stats.py.
 */
public final class Stats {

    private static final int[] WINDOWS = {30, 60, 180, 1800};
    private static final DateTimeFormatter TIME_FMT = DateTimeFormatter.ofPattern("HH:mm:ss")
            .withZone(ZoneId.systemDefault());

    private final Object lock = new Object();
    private final Integer yellowThresholdSeconds;
    private long sentTotal = 0;
    private long recvTotal = 0;
    private final Deque<Long> sentTimesMillis = new ArrayDeque<>();
    private final Deque<Long> recvTimesMillis = new ArrayDeque<>();
    private Long lastRecvMillis = null;
    private final Map<String, Boolean> connections = new LinkedHashMap<>();
    private final Map<String, Object> gauges = new LinkedHashMap<>();

    public Stats(Integer yellowThresholdSeconds) {
        this.yellowThresholdSeconds = yellowThresholdSeconds;
    }

    public void setConnection(String name, boolean connected) {
        synchronized (lock) {
            connections.put(name, connected);
        }
    }

    public void setGauge(String name, Object value) {
        synchronized (lock) {
            gauges.put(name, value);
        }
    }

    public void recordSent() {
        synchronized (lock) {
            sentTotal++;
            sentTimesMillis.addLast(System.currentTimeMillis());
            trim(sentTimesMillis);
        }
    }

    public void recordRecv() {
        synchronized (lock) {
            long now = System.currentTimeMillis();
            recvTotal++;
            recvTimesMillis.addLast(now);
            trim(recvTimesMillis);
            lastRecvMillis = now;
        }
    }

    private static void trim(Deque<Long> times) {
        long cutoff = System.currentTimeMillis() - maxWindowMillis();
        while (!times.isEmpty() && times.peekFirst() < cutoff) {
            times.pollFirst();
        }
    }

    private static long maxWindowMillis() {
        int max = 0;
        for (int w : WINDOWS) {
            max = Math.max(max, w);
        }
        return max * 1000L;
    }

    private static int countWithin(Deque<Long> times, int windowSeconds) {
        long cutoff = System.currentTimeMillis() - windowSeconds * 1000L;
        int count = 0;
        for (Long t : times) {
            if (t >= cutoff) {
                count++;
            }
        }
        return count;
    }

    public Map<String, Object> snapshot() {
        synchronized (lock) {
            Map<String, Object> result = new LinkedHashMap<>();
            result.put("sent_total", sentTotal);
            result.put("recv_total", recvTotal);
            for (int window : WINDOWS) {
                result.put("sent_" + window + "s", countWithin(sentTimesMillis, window));
                result.put("recv_" + window + "s", countWithin(recvTimesMillis, window));
            }

            if (lastRecvMillis != null) {
                double secondsSince = (System.currentTimeMillis() - lastRecvMillis) / 1000.0;
                result.put("seconds_since_last_recv", Math.round(secondsSince * 10) / 10.0);
                result.put("last_recv_datetime", TIME_FMT.format(Instant.ofEpochMilli(lastRecvMillis)));
            } else {
                result.put("seconds_since_last_recv", null);
                result.put("last_recv_datetime", null);
            }

            if (yellowThresholdSeconds != null) {
                result.put("yellow_threshold_seconds", yellowThresholdSeconds);
            }

            if (!connections.isEmpty()) {
                result.put("connections", new LinkedHashMap<>(connections));
            }

            if (!gauges.isEmpty()) {
                result.put("gauges", new LinkedHashMap<>(gauges));
            }

            return result;
        }
    }
}

File: claude_exp/routers/router_py/shared/stats.py
import threading
import time
from collections import deque
from datetime import datetime

_WINDOWS = (30, 60, 180, 1800)


class Stats:
    def __init__(self, yellow_threshold_seconds=None):
        self._lock = threading.Lock()
        self._yellow_threshold_seconds = yellow_threshold_seconds
        self._sent_total = 0
        self._recv_total = 0
        self._sent_times = deque()
        self._recv_times = deque()
        self._last_recv_time = None
        self._connections = {}
        self._gauges = {}

    def set_connection(self, name: str, connected: bool) -> None:
        with self._lock:
            self._connections[name] = bool(connected)

    def set_gauge(self, name: str, value) -> None:
        with self._lock:
            self._gauges[name] = value

    def record_sent(self) -> None:
        with self._lock:
            self._sent_total += 1
            self._sent_times.append(time.time())
            self._trim(self._sent_times)

    def record_recv(self) -> None:
        with self._lock:
            now = time.time()
            self._recv_total += 1
            self._recv_times.append(now)
            self._trim(self._recv_times)
            self._last_recv_time = now

    @staticmethod
    def _trim(times: deque) -> None:
        cutoff = time.time() - max(_WINDOWS)
        while times and times[0] < cutoff:
            times.popleft()

    @staticmethod
    def _count_within(times: deque, window_seconds: int) -> int:
        cutoff = time.time() - window_seconds
        count = 0
        for t in reversed(times):
            if t < cutoff:
                break
            count += 1
        return count

    def snapshot(self) -> dict:
        with self._lock:
            result = {"sent_total": self._sent_total, "recv_total": self._recv_total}
            for window in _WINDOWS:
                result[f"sent_{window}s"] = self._count_within(self._sent_times, window)
                result[f"recv_{window}s"] = self._count_within(self._recv_times, window)

            if self._last_recv_time is not None:
                result["seconds_since_last_recv"] = round(time.time() - self._last_recv_time, 1)
                result["last_recv_datetime"] = datetime.fromtimestamp(self._last_recv_time).strftime("%H:%M:%S")
            else:
                result["seconds_since_last_recv"] = None
                result["last_recv_datetime"] = None

            if self._yellow_threshold_seconds is not None:
                result["yellow_threshold_seconds"] = self._yellow_threshold_seconds

            if self._connections:
                result["connections"] = dict(self._connections)

            if self._gauges:
                result["gauges"] = dict(self._gauges)

            return result
